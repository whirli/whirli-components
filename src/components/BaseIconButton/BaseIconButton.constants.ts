export const PropKeys: Record<string, string> = {
  TAG: 'tag',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.TAG]: ['button', 'a'],
};
