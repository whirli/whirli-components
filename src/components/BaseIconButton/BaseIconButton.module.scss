@import '../../styles/Breakpoints.module.scss';

// default styles
.icon-button {
    background: none;
    border: none;
    box-shadow: none;
    color: inherit;
    cursor: pointer;
    display: inline-block;
}

.icon-button:disabled {
    cursor: not-allowed;
}


// breakpoint styles
$packageBreakpointStyles: ();