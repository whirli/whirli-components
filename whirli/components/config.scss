@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}