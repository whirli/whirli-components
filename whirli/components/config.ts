import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
