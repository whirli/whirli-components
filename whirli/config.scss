@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}