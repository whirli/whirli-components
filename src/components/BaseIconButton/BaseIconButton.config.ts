// @ts-ignore
import type { PropType } from 'vue';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BaseIconButton/BaseIconButton.config';

// Package config
import { PropKeys, PropValues } from './BaseIconButton.constants';
import type { PropTag } from './BaseIconButton.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.IS_DISABLED]: {
    required: false,
    type: Boolean as PropType<boolean>,
    default: false,
  },
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    validator: (value: PropTag): boolean => PropValues[PropKeys.TAG].includes(value),
    default: 'button',
  },
  ...config.props,
};
