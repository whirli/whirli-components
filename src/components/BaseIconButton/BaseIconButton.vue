<template>
  <component :is="props.tag" :class="[classes, disabledClasses]">
    <BaseIcon v-bind="$attrs" />
  </component>
</template>

<script setup lang="ts">
// Vue
import { computed, ComputedRef } from '@composition';

// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseIconButton/BaseIconButton.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseIconButton.config';
import { PropKeys } from '@whirli-components/components/BaseIconButton/BaseIconButton.constants';

// Components
import BaseIcon from '@whirli-components/components/BaseIcon/BaseIcon.vue';

// Composables
import useClasses from '@whirli-components/@use/class';

const ComponentStyles = ConfigStyles;

// @ts-ignore
const props = defineProps(ConfigProps);

// Classes
const { makeClasses } = useClasses();
const classes: string[] = [
  styles['icon-button'],
  ...makeClasses(ComponentStyles, ConfigProps, props, styles),
];
const disabledClasses: ComputedRef<string> = computed(() =>
  props[PropKeys.IS_DISABLED] ? styles['is-disabled'] : ''
);
</script>
