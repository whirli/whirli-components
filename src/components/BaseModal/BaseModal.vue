<template>
  <BaseOverlay :state="props.state">
    <div
      ref="modal"
      :class="wrapperClasses"
      aria-modal="true"
      :aria-labelledby="uniqueId"
      role="dialog"
      :aria-hidden="!isOpen"
    >
      <BaseSpacer size="3xl" />
      <div :id="uniqueId" :class="titleClasses">
        <slot name="title" />
      </div>
      <div :class="contentClasses">
        <slot name="content" />
      </div>
      <BaseSpacer size="3xl" />
      <div :class="styles['modal__close-icon']">
        <BaseIconButton ref="modalClose" icon="Cross" @click="emit('modal:close')" aria-label="Close modal" />
      </div>
    </div>
  </BaseOverlay>
</template>

<script setup lang="ts">
// Vue
import { defineEmits, watch, ref, computed, Ref, ComputedRef } from '@composition';

// Components
import BaseOverlay from '@whirli-components/components/BaseOverlay/BaseOverlay.vue';
import BaseSpacer from '@whirli-components/components/BaseSpacer/BaseSpacer.vue';
import BaseIconButton from '@whirli-components/components/BaseIconButton/BaseIconButton.vue';

// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseModal/BaseModal.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseModal.config';

// Composables
import useComponent from '@whirli-components/@use/component';
import useFocusTrap from '@whirli-components/@use/focusTrap';
import useClasses from '@whirli-components/@use/class';

const { uniqueId } = useComponent();
const { initFocusTrap, destroyFocusTrap } = useFocusTrap();

const ComponentStyles = ConfigStyles;

// @ts-ignore
const props = defineProps(ConfigProps);
// @ts-ignore
const emit = defineEmits(['modal:close', 'modal:open']);

// Classes
const { makeClasses } = useClasses();
const wrapperClasses = [styles.modal, ...makeClasses(ComponentStyles, ConfigProps, props, styles)];
const titleClasses = [styles['modal__title']];
const contentClasses = [styles['modal__content']];

const modal: Ref<HTMLElement> = ref(null as unknown as HTMLElement);
const modalClose: Ref<HTMLElement> = ref(null as unknown as HTMLElement);

const isOpen: ComputedRef<boolean> = computed(() => props.state === 'open');

watch(
  () => props.state,
  (newState: string) => {
    if (newState === 'open') {
      initFocusTrap(modal.value);
    } else {
      destroyFocusTrap(modal.value);
    }
  }
);
</script>
