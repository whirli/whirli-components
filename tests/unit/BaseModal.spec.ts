import { mount } from '@vue/test-utils';

import { propReturnsClass } from '../helpers';

import BaseModal from '@/components/BaseModal/BaseModal.vue';
import BaseIconButton from '@/components/BaseIconButton/BaseIconButton.vue';

import { PropKeys } from '@/components/BaseModal/BaseModal.constants';
import { ConfigStyles } from '@/components/BaseModal/BaseModal.config';

let wrapper: any;
const text = 'Hello World';

describe('BaseModal.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = mount(BaseModal, {
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the correct data', () => {
    expect(wrapper.text()).toBe(text);
  });

  it('renders the close button', () => {
    expect(wrapper.findComponent(BaseIconButton).exists()).toBeTruthy();
  });

  it('close button emits close event', async () => {
    wrapper.findComponent(BaseIconButton).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('item:click'));
  });

  it('is the correct color', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.COLOR, 'color-default')).toBe(true);
  });

  it('is the correct size', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SIZE, 'size-default')).toBe(true);
  });
});
