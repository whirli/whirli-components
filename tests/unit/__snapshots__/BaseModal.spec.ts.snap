// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModal.vue - Single classes renders the component 1`] = `
<div
  class=""
>
  <div
    class=""
  >
    
    <div
      aria-hidden="true"
      aria-labelledby="3c4f051b-6e6f-4f27-a77b-12bfd51958c7"
      aria-modal="true"
      class=""
      role="dialog"
    >
      <div
        class=""
      />
      <div
        class=""
        id="3c4f051b-6e6f-4f27-a77b-12bfd51958c7"
      >
        
        Modal title
        
      </div>
      <div
        class=""
      >
        
        Hello World
        
      </div>
      <div
        class=""
      />
      <div
        class=""
      >
        <button
          aria-label="Close modal"
          class=""
          icon="Cross"
        >
          <svg
            aria-label="Close modal"
            class=""
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M21.2 4.2l-1.4-1.4-7.8 7.8-7.8-7.8-1.4 1.4 7.8 7.8-7.8 7.8 1.4 1.4 7.8-7.8 7.8 7.8 1.4-1.4-7.8-7.8z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
    
  </div>
</div>
`;
